File: Assignment_1/app/src/main/java/com/kaushiknsanji/learndagger/MyApplication.java
package com.kaushiknsanji.learndagger;

import android.app.Application;

import com.kaushiknsanji.learndagger.data.local.DatabaseService;
import com.kaushiknsanji.learndagger.data.remote.NetworkService;
import com.kaushiknsanji.learndagger.di.component.ApplicationComponent;
import com.kaushiknsanji.learndagger.di.component.DaggerApplicationComponent;
import com.kaushiknsanji.learndagger.di.module.ApplicationModule;
import com.kaushiknsanji.learndagger.utils.NetworkHelper;

import javax.inject.Inject;

/**
 * {@link Application} subclass for exposing {@link ApplicationComponent}
 */
public class MyApplication extends Application {

    @Inject
    NetworkService networkService;
    @Inject
    DatabaseService databaseService;
    @Inject
    NetworkHelper networkHelper;
    private ApplicationComponent applicationComponent;

    @Override
    public void onCreate() {
        super.onCreate();

        getApplicationComponent().inject(this);
    }

    public ApplicationComponent getApplicationComponent() {
        if (applicationComponent == null) {
            applicationComponent = DaggerApplicationComponent
                    .builder()
                    .applicationModule(new ApplicationModule(this))
                    .build();
        }
        return applicationComponent;
    }

}

File: Assignment_2/app/src/main/java/com/kaushiknsanji/learndagger/MyApplication.kt
package com.kaushiknsanji.learndagger

import android.app.Application
import com.kaushiknsanji.learndagger.data.local.DatabaseService
import com.kaushiknsanji.learndagger.data.remote.NetworkService
import com.kaushiknsanji.learndagger.di.component.ApplicationComponent
import com.kaushiknsanji.learndagger.di.component.DaggerApplicationComponent
import com.kaushiknsanji.learndagger.di.module.ApplicationModule
import com.kaushiknsanji.learndagger.utils.NetworkHelper
import javax.inject.Inject

/**
 * [Application] subclass for exposing [ApplicationComponent]
 *
 * @author Kaushik N Sanji
 */
class MyApplication : Application() {

    @Inject
    lateinit var networkService: NetworkService

    @Inject
    lateinit var databaseService: DatabaseService

    @Inject
    lateinit var networkHelper: NetworkHelper

    //[ApplicationComponent] built lazily and only once
    val applicationComponent: ApplicationComponent by lazy {
        DaggerApplicationComponent.builder()
            .applicationModule(ApplicationModule(this))
            .build()
    }

    override fun onCreate() {
        super.onCreate()

        applicationComponent.inject(this)
    }

}